File: app/src/main/java/com/moneycollect/example_java/fragment/SaveWithPaymentCardFragment.java
package com.moneycollect.example_java.fragment;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.activity.result.ActivityResult;
import androidx.activity.result.ActivityResultCallback;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;

import com.moneycollect.android.model.MoneyCollectCardListViewParams;
import com.moneycollect.android.model.enumeration.MoneyCollectPaymentModel;
import com.moneycollect.android.model.request.RequestConfirmPayment;
import com.moneycollect.android.model.request.RequestCreatePayment;
import com.moneycollect.android.model.response.Payment;
import com.moneycollect.android.model.response.PaymentMethod;
import com.moneycollect.android.net.net.ApiResultCallback;
import com.moneycollect.android.ui.imp.MoneyCollectResultBackInterface;
import com.moneycollect.android.ui.view.MoneyCollectCardListView;
import com.moneycollect.example_java.R;
import com.moneycollect.example_java.databinding.FragmentSaveWithPaymentLayoutBinding;
import com.moneycollect.example_java.BaseExampleFragment;
import com.moneycollect.example_java.Constant;
import com.moneycollect.example_java.TestRequestData;
import com.moneycollect.example_java.activity.PayCardActivity;
import com.moneycollect.example_java.activity.ValidationWebActivity;

import org.jetbrains.annotations.NotNull;

import java.math.BigInteger;
import java.util.ArrayList;

/**
 * {@link SaveWithPaymentCardFragment}
 * Show the payment list,offer users pay the payment, also you can add a new card to pay jump {@link AddWithPaymentFragment}
 */
public class SaveWithPaymentCardFragment extends BaseExampleFragment implements View.OnClickListener {

    private FragmentSaveWithPaymentLayoutBinding viewBinding;

    /***  [MoneyCollectCardListView]*/
    MoneyCollectCardListView cardListLayout;

    /***  [MoneyCollectResultBackInterface]*/
    private MoneyCollectResultBackInterface moneyCollectResultBackInterface;

    // PaymentModel (PAY)
    private MoneyCollectPaymentModel currentModel = MoneyCollectPaymentModel.PAY;

    //RequestCreatePayment params for pay
    private RequestCreatePayment currentRequestCreatePayment;
    //RequestConfirmPayment params for pay
    private RequestConfirmPayment currentRequestConfirmPayment;

    //customerId for pay
    private String customerId;

    //groupList params
    private ArrayList<PaymentMethod> groupList = new ArrayList<>();
    //childList params
    private ArrayList<ArrayList<PaymentMethod>> childList = new ArrayList<>();

    //loading active
    private boolean isRequestLoading = false;

    private static String ADD_PAYMENT = "AddWithPaymentFragment";

    @androidx.annotation.Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @androidx.annotation.Nullable ViewGroup container, @androidx.annotation.Nullable Bundle savedInstanceState) {
        viewBinding = FragmentSaveWithPaymentLayoutBinding.inflate(getLayoutInflater(), container, false);
        initUi();
        return viewBinding.getRoot();
    }

    private void initUi() {
        if (getArguments() != null) {
            currentModel = (MoneyCollectPaymentModel) getArguments().getSerializable(Constant.CURRENT_PAYMENT_MODEL);
            currentRequestCreatePayment = getArguments().getParcelable(Constant.CREATE_PAYMENT_REQUEST_TAG);
            currentRequestConfirmPayment = getArguments().getParcelable(Constant.CONFIRM_PAYMENT_REQUEST_TAG);
            customerId = getArguments().getString(Constant.CUSTOMER_ID_TAG);
        }
        cardListLayout = viewBinding.mcCardList;
        moneyCollectResultBackInterface = cardListLayout.gainMoneyCollectResultBackInterface();
        MoneyCollectCardListViewParams moneyCollectCardListViewParams = new MoneyCollectCardListViewParams.Builder()
                .activity(getActivity())
                .moneyCollectPaymentModel(currentModel)
                .addClickListener(this)
                .build();
        cardListLayout.setMoneyCollectCardListViewParams(moneyCollectCardListViewParams);

        cardListLayout.getPaymentMethodParamsLiveData().observe(getActivity(),
                result -> {
                    if (result != null) {
                        createPayment(result);
                    } else {
                        moneyCollectResultBackInterface.failExceptionBack(getString(R.string.payment_method_list_empty_str));
                    }
                }
        );
        selectAllPaymentMethods();
    }


    /**
     * create payment
     */
    private void createPayment(PaymentMethod paymentMethod) {
        if (currentRequestCreatePayment == null) {
            moneyCollectResultBackInterface.failExceptionBack(getString(R.string.request_create_payment_empty_str));
            return;
        }
        if (TextUtils.isEmpty(currentRequestCreatePayment.getCustomerId())) {
            moneyCollectResultBackInterface.failExceptionBack(getString(R.string.customer_id_empty_str));
            return;
        }
        RequestCreatePayment requestCreatePayment = new RequestCreatePayment(
                currentRequestCreatePayment.getAutomaticPaymentMethods(),
                currentRequestCreatePayment.getAmount(),
                false,
                currentRequestCreatePayment.getConfirmationMethod(),
                currentRequestCreatePayment.getCurrency(),
                currentRequestCreatePayment.getCustomerId(),
                currentRequestCreatePayment.getDescription(),
                currentRequestCreatePayment.getFromChannel(),
                currentRequestCreatePayment.getIp(),
                currentRequestCreatePayment.getNotifyUrl(),
                currentRequestCreatePayment.getOrderNo(),
                paymentMethod.id,
                TestRequestData.Companion.getPaymentMethodTypes(),
                currentRequestCreatePayment.getPreAuth(),
                currentRequestCreatePayment.getReceiptEmail(),
                currentRequestCreatePayment.getReturnUrl(),
                currentRequestCreatePayment.getSetupFutureUsage(),
                new RequestCreatePayment.Shipping(
                        currentRequestCreatePayment.getShipping().address,
                        currentRequestCreatePayment.getShipping().firstName,
                        currentRequestCreatePayment.getShipping().lastName,
                        currentRequestCreatePayment.getShipping().phone
                ),
                currentRequestCreatePayment.getLineItems(),
                currentRequestCreatePayment.getStatementDescriptor(),
                currentRequestCreatePayment.getStatementDescriptorSuffix(),
                currentRequestCreatePayment.getUserAgent(),
                currentRequestCreatePayment.getWebsite()
        );

        if (requestCreatePayment.getConfirmationMethod() == RequestCreatePayment.ConfirmationMethod.Automatic) {
            if (TextUtils.isEmpty(requestCreatePayment.getPaymentMethod())) {
                moneyCollectResultBackInterface.failExceptionBack(getString(R.string.payment_method_empty_str));
                return;
            }
        }
        moneyCollect.createPayment(requestCreatePayment, new ApiResultCallback<Payment>() {
            @Override
            public void onSuccess(@NotNull Payment result) {
                confirmPayment(result);
            }

            @Override
            public void onError(@NotNull Exception e) {
                moneyCollectResultBackInterface.failExceptionBack(e.getMessage());
            }
        });

    }

    /**
     * confirm payment
     */
    private void confirmPayment(Payment payment) {
        if (currentRequestConfirmPayment == null) {
            moneyCollectResultBackInterface.failExceptionBack(getString(R.string.request_confirm_payment_empty_str));
            return;
        }
        if (TextUtils.isEmpty(payment.getPaymentMethod())) {
            moneyCollectResultBackInterface.failExceptionBack(getString(R.string.payment_method_empty_str));
            return;
        }
        if (TextUtils.isEmpty(payment.getId())) {
            moneyCollectResultBackInterface.failExceptionBack(getString(R.string.payment_id_empty_str));
            return;
        }
        if (TextUtils.isEmpty(payment.getClientSecret())) {
            moneyCollectResultBackInterface.failExceptionBack(getString(R.string.payment_client_secret_empty_str));
            return;
        }
        RequestConfirmPayment requestConfirmPayment = new RequestConfirmPayment(
                BigInteger.valueOf(payment.getAmount()),
                payment.getCurrency(),
                payment.getId(),
                payment.getIp(),
                payment.getNotifyUrl(),
                payment.getPaymentMethod(),
                payment.getReceiptEmail(),
                payment.getReturnUrl(),
                payment.getSetupFutureUsage(),
                new RequestConfirmPayment.Shipping(
                        currentRequestConfirmPayment.getShipping().address,
                        currentRequestConfirmPayment.getShipping().firstName,
                        currentRequestConfirmPayment.getShipping().lastName,
                        currentRequestConfirmPayment.getShipping().phone
                ), currentRequestConfirmPayment.getWebsite()
        );

        moneyCollect.confirmPayment(requestConfirmPayment, payment.getClientSecret(), new ApiResultCallback<Payment>() {
                    @Override
                    public void onSuccess(@NotNull Payment result) {
                        if (result==null){
                            moneyCollectResultBackInterface.paymentConfirmResultBack(false, Constant.PAYMENT_PENDING_MESSAGE);
                            return;
                        }
                        if (result.getNextAction() != null) {
                            String redirectToUrl = result.getNextAction().redirectToUrl;
                            if (!TextUtils.isEmpty(redirectToUrl)) {
                                Intent intent = new Intent(getActivity(), ValidationWebActivity.class);
                                intent.putExtra(Constant.VALIDATION_PARAM_URL, redirectToUrl);
                                intent.putExtra(Constant.VALIDATION_PAYMENT_ID, result.getId());
                                intent.putExtra(Constant.VALIDATION_PAYMENT_CLIENTSECRET, result.getClientSecret());
                                startActivityLauncher.launch(intent);
                            } else {
                                moneyCollectResultBackInterface.paymentConfirmResultBack(false, Constant.PAYMENT_PENDING_MESSAGE);
                            }
                        } else {
                            String status = result.getStatus();
                            if (status != null) {
                                switch (status) {
                                    case Constant.PAYMENT_SUCCEEDED:
                                        Intent intent = new Intent();
                                        intent.putExtra(Constant.PAYMENT_RESULT_PAYMENT, result);
                                        if (getActivity()!=null) {
                                            getActivity().setResult(Constant.PAYMENT_RESULT_CODE, intent);
                                        }
                                        moneyCollectResultBackInterface.paymentConfirmResultBack(true, "");
                                        break;
                                    case Constant.PAYMENT_FAILED:
                                        moneyCollectResultBackInterface.paymentConfirmResultBack(false, result.getErrorMessage());
                                        break;
                                    case Constant.PAYMENT_UN_CAPTURED:
                                        moneyCollectResultBackInterface.paymentConfirmResultBack(false, Constant.PAYMENT_UN_CAPTURED_MESSAGE);
                                        break;
                                    case Constant.PAYMENT_PENDING:
                                        moneyCollectResultBackInterface.paymentConfirmResultBack(false, Constant.PAYMENT_PENDING_MESSAGE);
                                        break;
                                    case Constant.PAYMENT_CANCELED:
                                        moneyCollectResultBackInterface.paymentConfirmResultBack(false, Constant.PAYMENT_CANCELED_MESSAGE);
                                        break;
                                    default:
                                        moneyCollectResultBackInterface.paymentConfirmResultBack(false, Constant.PAYMENT_PENDING_MESSAGE);
                                        break;
                                }
                            } else {
                                moneyCollectResultBackInterface.paymentConfirmResultBack(false, Constant.PAYMENT_PENDING_MESSAGE);
                            }
                        }
                    }

                    @Override
                    public void onError(@NotNull Exception e) {
                        moneyCollectResultBackInterface.failExceptionBack(e.getMessage());
                    }
                }
        );

    }

    private ActivityResultLauncher<Intent> startActivityLauncher = registerForActivityResult(
            new ActivityResultContracts.StartActivityForResult(), new ActivityResultCallback<ActivityResult>() {
                @Override
                public void onActivityResult(ActivityResult result) {
                    if (result == null) return;
                    if (result.getResultCode() == Constant.WEB_RESULT_CODE) {
                        String resultStr = "";
                        Intent dataIntent = result.getData();
                        Payment payment = null;
                        if (dataIntent != null) {
                            resultStr = dataIntent.getStringExtra(Constant.WEB_RESULT_TAG);
                            payment = dataIntent.getParcelableExtra(Constant.PAYMENT_RESULT_PAYMENT);
                        }
                        if (TextUtils.isEmpty(resultStr)) {
                            Intent intent = new Intent();
                            intent.putExtra(Constant.WEB_RESULT_TAG, resultStr);
                            intent.putExtra(Constant.PAYMENT_RESULT_PAYMENT, payment);
                            if (getActivity()!=null) {
                                getActivity().setResult(Constant.PAYMENT_RESULT_CODE, intent);
                            }
                            moneyCollectResultBackInterface.paymentConfirmResultBack(true, "");
                        } else {
                            moneyCollectResultBackInterface.failExceptionBack(resultStr);
                        }
                    }
                }
            }
    );


    /**
     * select all paymentMethods
     */
    private void selectAllPaymentMethods() {
        if (TextUtils.isEmpty(customerId)) {
            cardListLayout.setVisibility(View.VISIBLE);
            cardListLayout.setPaymentButtonAndFootViewStatus(false);
            return;
        }
        isRequestLoading = true;
        showLoadingDialog();
        moneyCollect.selectAllPaymentMethods(customerId, new ApiResultCallback<Object>() {
            @Override
            public void onSuccess(@NotNull Object result) {
                dismissLoadingDialog();
                cardListLayout.setVisibility(View.VISIBLE);
                if (result instanceof ArrayList) {
                    if (((ArrayList) result).size() > 0) {
                        childList.add((ArrayList<PaymentMethod>) result);
                        if (childList.size() > 0) {
                            groupList.add(childList.get(0).get(0));
                            cardListLayout.setDataList(groupList, childList);
                            cardListLayout.setPaymentButtonAndFootViewStatus(true);
                        } else {
                            cardListLayout.setPaymentButtonAndFootViewStatus(false);
                        }
                    } else {
                        cardListLayout.setPaymentButtonAndFootViewStatus(false);
                    }
                } else {
                    cardListLayout.setPaymentButtonAndFootViewStatus(false);
                }
                isRequestLoading = false;
            }

            @Override
            public void onError(@NotNull Exception e) {
                dismissLoadingDialog();
                cardListLayout.setVisibility(View.VISIBLE);
                cardListLayout.setPaymentButtonAndFootViewStatus(false);
                isRequestLoading = false;
            }
        });

    }

    /**
     * click event
     */
    @Override
    public void onClick(View view) {
        Activity activity = getActivity();
        if (view != null && activity!=null) {
            if (view.getId() == cardListLayout.getToolbarBackIcon().getId() && !cardListLayout.isRequestLoading() && !isRequestLoading) {
                activity.finish();
            } else if (view.getId() == cardListLayout.getChildFootView().getId() && !cardListLayout.isRequestLoading() && !isRequestLoading) {
                if (activity instanceof PayCardActivity) {
                    ((PayCardActivity)activity).switchContent(ADD_PAYMENT);
                }
            }
        }
    }
}

File: app/src/main/java/com/moneycollect/example/fragment/SaveWithPaymentCardFragment.kt
package com.moneycollect.example.fragment

import android.content.Intent
import android.os.Bundle
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import com.moneycollect.android.model.MoneyCollectCardListViewParams
import com.moneycollect.android.model.enumeration.MoneyCollectPaymentModel
import com.moneycollect.android.model.request.RequestConfirmPayment
import com.moneycollect.android.model.request.RequestCreatePayment
import com.moneycollect.android.model.response.Payment
import com.moneycollect.android.model.response.PaymentMethod
import com.moneycollect.android.net.net.ApiResultCallback
import com.moneycollect.android.ui.imp.MoneyCollectResultBackInterface
import com.moneycollect.android.ui.view.MoneyCollectCardListView
import com.moneycollect.example.BaseExampleFragment
import com.moneycollect.example.Constant
import com.moneycollect.example.R
import com.moneycollect.example.TestRequestData
import com.moneycollect.example.activity.PayCardActivity
import com.moneycollect.example.activity.ValidationWebActivity
import com.moneycollect.example.databinding.FragmentSaveWithPaymentLayoutBinding
import java.util.*

/**
 * [SaveWithPaymentCardFragment]
 * Show the payment list,offer users pay the payment, also you can add a new card to pay jump [AddWithPaymentFragment]
 *
 */
class SaveWithPaymentCardFragment : BaseExampleFragment(),View.OnClickListener{

    private var viewBinding: FragmentSaveWithPaymentLayoutBinding?=null

    /***  [MoneyCollectCardListView]*/
    var cardListLayout: MoneyCollectCardListView?=null

    /***  [MoneyCollectResultBackInterface]*/
    private var  moneyCollectResultBackInterface: MoneyCollectResultBackInterface?=null

    // PaymentModel (PAY)
    private var currentModel:MoneyCollectPaymentModel=MoneyCollectPaymentModel.PAY

    //customerId
    var customerId: String?=null

    //RequestCreatePayment params for pay
    private var currentRequestCreatePayment:RequestCreatePayment?=null
    //RequestConfirmPayment params for pay
    private var currentRequestConfirmPayment:RequestConfirmPayment?=null

    //groupList params
    private var groupList: ArrayList<PaymentMethod> = ArrayList<PaymentMethod>()
    //childList params
    private var childList: ArrayList<ArrayList<PaymentMethod>> = ArrayList<ArrayList<PaymentMethod>>()

    //loading active
    private var isRequestLoading=false

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View? {
        viewBinding = FragmentSaveWithPaymentLayoutBinding.inflate(layoutInflater, container, false)
        initUi()
        return viewBinding?.root
    }

    private fun initUi() {
        currentModel= arguments?.getSerializable(Constant.CURRENT_PAYMENT_MODEL) as MoneyCollectPaymentModel
        currentRequestCreatePayment= arguments?.getParcelable(Constant.CREATE_PAYMENT_REQUEST_TAG)
        currentRequestConfirmPayment= arguments?.getParcelable(Constant.CONFIRM_PAYMENT_REQUEST_TAG)
        customerId= arguments?.getString(Constant.CUSTOMER_ID_TAG)

        cardListLayout= viewBinding?.mcCardList
        moneyCollectResultBackInterface=cardListLayout?.gainMoneyCollectResultBackInterface()
        val moneyCollectCardListViewParams = MoneyCollectCardListViewParams.Builder()
            .activity(activity)
            .moneyCollectPaymentModel(currentModel)
            .addClickListener(this)
            .build()
        cardListLayout?.setMoneyCollectCardListViewParams(moneyCollectCardListViewParams)

        //Listening to pay button return data, further request
        activity?.let {
            cardListLayout?.paymentMethodParamsLiveData?.observe(it,
                { result ->
                    result.apply {
                        if (result != null) {
                            createPayment(result)
                        } else {
                            moneyCollectResultBackInterface?.failExceptionBack(getString(R.string.payment_method_list_empty_str))
                        }
                    }
                }
            )
        }
        selectAllPaymentMethods()
    }


    /**
     * create payment
     */
    private fun createPayment(paymentMethod: PaymentMethod) {
        if (currentRequestCreatePayment == null){
            moneyCollectResultBackInterface?.failExceptionBack(getString(R.string.request_create_payment_empty_str))
            return
        }
        if (TextUtils.isEmpty(currentRequestCreatePayment?.customerId)) {
            moneyCollectResultBackInterface?.failExceptionBack(getString(R.string.customer_id_empty_str))
            return
        }
        currentRequestCreatePayment?.let {
            val requestCreatePayment = RequestCreatePayment(
                automaticPaymentMethods = it.automaticPaymentMethods,
                amount = it.amount,
                confirm = false,
                confirmationMethod = it.confirmationMethod,
                currency = it.currency,
                customerId = it.customerId,
                description = it.description,
                fromChannel = it.fromChannel,
                ip = it.ip,
                lineItems = it.lineItems,
                notifyUrl = it.notifyUrl,
                orderNo = it.orderNo,
                paymentMethod = paymentMethod.id,
                paymentMethodTypes = TestRequestData.paymentMethodTypes,
                preAuth = it.preAuth,
                receiptEmail = it.receiptEmail,
                returnUrl = it.returnUrl,
                setupFutureUsage = it.setupFutureUsage,
                shipping = RequestCreatePayment.Shipping(
                    address = it.shipping?.address,
                    firstName = it.shipping?.firstName,
                    lastName = it.shipping?.lastName,
                    phone = it.shipping?.phone
                ),
                statementDescriptor = it.statementDescriptor,
                statementDescriptorSuffix = it.statementDescriptorSuffix,
                userAgent = it.userAgent,
                website = it.website
            )
            if (requestCreatePayment.confirmationMethod == RequestCreatePayment.ConfirmationMethod.Automatic) {
                if (TextUtils.isEmpty(requestCreatePayment.paymentMethod)) {
                    moneyCollectResultBackInterface?.failExceptionBack(getString(R.string.payment_method_empty_str))
                    return
                }
            }
            moneyCollect.createPayment(requestCreatePayment,
                object : ApiResultCallback<Payment> {
                    override fun onSuccess(result: Payment) {
                        confirmPayment(result, paymentMethod)
                    }

                    override fun onError(e: Exception) {
                        moneyCollectResultBackInterface?.failExceptionBack(e.message)
                    }
                })
        }
    }

    /**
     * confirm payment
     */
    private fun confirmPayment(payment: Payment, paymentMethod: PaymentMethod) {
        if (currentRequestConfirmPayment == null){
            moneyCollectResultBackInterface?.failExceptionBack(getString(R.string.request_confirm_payment_empty_str))
            return
        }
        if (TextUtils.isEmpty(payment.paymentMethod)) {
            moneyCollectResultBackInterface?.failExceptionBack(getString(R.string.payment_method_empty_str))
            return
        }
        if (TextUtils.isEmpty(payment.id)) {
            moneyCollectResultBackInterface?.failExceptionBack(getString(R.string.payment_id_empty_str))
            return
        }
        if (TextUtils.isEmpty(payment.clientSecret)) {
            moneyCollectResultBackInterface?.failExceptionBack(getString(R.string.payment_client_secret_empty_str))
            return
        }
        currentRequestConfirmPayment?.let {
            val requestConfirmPayment = RequestConfirmPayment(
                amount = payment.amount?.toBigInteger(),
                currency = payment.currency,
                id = payment.id,
                ip = payment.ip,
                notifyUrl = payment.notifyUrl,
                paymentMethod = payment.paymentMethod,
                receiptEmail = payment.receiptEmail,
                returnUrl = payment.returnUrl,
                setupFutureUsage = payment.setupFutureUsage,
                shipping = RequestConfirmPayment.Shipping(
                    address = it.shipping?.address,
                    firstName = it.shipping?.firstName,
                    lastName = it.shipping?.lastName,
                    phone = it.shipping?.phone
                ),
                website = it.website
            )

            moneyCollect.confirmPayment(requestConfirmPayment, payment.clientSecret,
                object : ApiResultCallback<Payment> {
                    override fun onSuccess(result: Payment) {
                        //if nextAction object is not null and redirectToUrl address is not null, further 3 d verification
                        if (result.nextAction != null) {
                            if (!TextUtils.isEmpty(result.nextAction?.type)) {
                                var redirectToUrl=result.nextAction?.redirectToUrl
                                if (result.nextAction?.type?.equals(TestRequestData.weChatPayNextActionType) == true) {
                                    redirectToUrl=result.nextAction?.wechatPayH5?.redirectToUrl
                                }
                                if (!TextUtils.isEmpty(redirectToUrl)) {
                                    val intent = Intent(activity, ValidationWebActivity::class.java)
                                    intent.putExtra(Constant.VALIDATION_PARAM_URL, redirectToUrl)
                                    intent.putExtra(Constant.VALIDATION_PAYMENT_ID, result.id)
                                    intent.putExtra(Constant.VALIDATION_PAYMENT_CLIENTSECRET, result.clientSecret)
                                    startActivityLauncher.launch(intent)
                                } else {
                                    moneyCollectResultBackInterface?.paymentConfirmResultBack(false,
                                        Constant.PAYMENT_PENDING_MESSAGE)
                                }
                            }else {
                                moneyCollectResultBackInterface?.paymentConfirmResultBack(false,
                                    Constant.PAYMENT_PENDING_MESSAGE)
                            }
                        } else {
                            //Need to deal with the state has succeeded, uncaptured, pending, failed, canceled
                            when (result.status) {
                                Constant.PAYMENT_SUCCEEDED -> {
                                    var intent =  Intent()
                                    intent.putExtra(Constant.PAYMENT_RESULT_PAYMENT, result)
                                    activity?.setResult(Constant.PAYMENT_RESULT_CODE,intent)
                                    moneyCollectResultBackInterface?.paymentConfirmResultBack(true,
                                        "")
                                }
                                Constant.PAYMENT_FAILED -> {
                                    moneyCollectResultBackInterface?.paymentConfirmResultBack(false,
                                        result.errorMessage)
                                }
                                Constant.PAYMENT_UN_CAPTURED -> {
                                    moneyCollectResultBackInterface?.paymentConfirmResultBack(false,
                                        Constant.PAYMENT_UN_CAPTURED_MESSAGE)
                                }
                                Constant.PAYMENT_PENDING -> {
                                    moneyCollectResultBackInterface?.paymentConfirmResultBack(false,
                                        Constant.PAYMENT_PENDING_MESSAGE)
                                }
                                Constant.PAYMENT_CANCELED -> {
                                    moneyCollectResultBackInterface?.paymentConfirmResultBack(false,
                                        Constant.PAYMENT_CANCELED_MESSAGE)
                                }
                                else -> {
                                    moneyCollectResultBackInterface?.paymentConfirmResultBack(false,
                                        Constant.PAYMENT_PENDING_MESSAGE)
                                }
                            }
                        }
                    }

                    override fun onError(e: Exception) {
                        moneyCollectResultBackInterface?.failExceptionBack(e.message)
                    }
                })
        }
    }

    private val startActivityLauncher: ActivityResultLauncher<Intent> =
        registerForActivityResult(ActivityResultContracts.StartActivityForResult()) {
            if (it.resultCode == Constant.WEB_RESULT_CODE) {
                var resultStr = it.data?.getStringExtra(Constant.WEB_RESULT_TAG)
                var payment =
                    it.data?.getParcelableExtra<Payment>(Constant.PAYMENT_RESULT_PAYMENT)
                if (resultStr != null) {
                    if (TextUtils.isEmpty(resultStr)){
                        var intent =  Intent()
                        intent.putExtra(Constant.WEB_RESULT_TAG, resultStr)
                        intent.putExtra(Constant.PAYMENT_RESULT_PAYMENT, payment)
                        activity?.setResult(Constant.PAYMENT_RESULT_CODE,intent)
                        moneyCollectResultBackInterface?.paymentConfirmResultBack(true, "")
                    }else{
                        moneyCollectResultBackInterface?.failExceptionBack(resultStr)
                    }
                }
            }
        }

    /**
     * select all paymentMethods
     */
    private fun selectAllPaymentMethods() {
        if (TextUtils.isEmpty(customerId)){
            cardListLayout?.visibility = View.VISIBLE
            cardListLayout?.setPaymentButtonAndFootViewStatus(false)
            return
        }
        customerId?.let {
            showLoadingDialog()
            isRequestLoading=true
            moneyCollect.selectAllPaymentMethods(it,
                object : ApiResultCallback<Any> {
                    override fun onSuccess(result: Any) {
                        dismissLoadingDialog()
                        cardListLayout?.visibility = View.VISIBLE
                        when (result) {
                            !is ArrayList<*> -> {
                                cardListLayout?.setPaymentButtonAndFootViewStatus(false)
                            }
                            else -> {
                                when {
                                    result.isEmpty() -> {
                                        cardListLayout?.setPaymentButtonAndFootViewStatus(false)
                                    }
                                    else -> {
                                        childList.add(result as ArrayList<PaymentMethod>)
                                        when {
                                            childList.isNotEmpty() -> {
                                                groupList.add(childList[0][0])
                                                cardListLayout?.setDataList(groupList, childList)
                                                cardListLayout?.setPaymentButtonAndFootViewStatus(
                                                    true)
                                            }
                                            else -> {
                                                cardListLayout?.setPaymentButtonAndFootViewStatus(
                                                    false)
                                            }
                                        }
                                    }
                                }
                            }
                        }
                        isRequestLoading = false
                    }

                    override fun onError(e: Exception) {
                        dismissLoadingDialog()
                        cardListLayout?.visibility = View.VISIBLE
                        cardListLayout?.setPaymentButtonAndFootViewStatus(false)
                        isRequestLoading = false
                    }
                })
        }
    }



    companion object{
        const val ADD_PAYMENT:String="AddWithPaymentFragment" //add fragment tag
    }

    /**
     * click event
     */
    override fun onClick(view: View?) {
        if (view!=null) {
            if (view.id == cardListLayout?.getToolbarBackIcon()?.id && cardListLayout?.isRequestLoading==false && !isRequestLoading) {
                activity?.finish()
            }else  if (view.id == cardListLayout?.getChildFootView()?.id && cardListLayout?.isRequestLoading==false && !isRequestLoading) {
                activity?.let {
                    (it as? PayCardActivity)?.switchContent(ADD_PAYMENT)
                }
            }
        }
    }
}